File: android/app/src/main/java/com/lilru/liftr/ui/tasks/WeeklyTasksHomeBadgeState.kt
package com.lilru.liftr.ui.tasks

import android.content.Context
import androidx.compose.ui.graphics.Color
import com.lilru.liftr.data.BackendContracts
import com.lilru.liftr.prefs.LiftrPreferences
import io.github.jan.supabase.postgrest.postgrest
import org.json.JSONObject
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class WeeklyTasksHomeSummary(
    @SerialName("week_start") val weekStart: String,
    @SerialName("accepted_count") val acceptedCount: Int = 0,
    @SerialName("accept_slots_remaining") val acceptSlotsRemaining: Int = 0,
    @SerialName("available_count") val availableCount: Int = 0,
    @SerialName("completed_count_this_week") val completedCountThisWeek: Int = 0,
) {
    val hasPickableTasks: Boolean
        get() = acceptSlotsRemaining > 0 && availableCount > 0
}

object WeeklyTasksHomeBadgeState {
    fun hasUnseenCompletion(context: Context, summary: WeeklyTasksHomeSummary): Boolean {
        val lastSeen = effectiveLastSeenCount(context, summary)
        return summary.completedCountThisWeek > lastSeen
    }

    private fun effectiveLastSeenCount(context: Context, summary: WeeklyTasksHomeSummary): Int {
        val storedWeek = LiftrPreferences.weeklyTasksLastSeenWeekStart(context)
        if (storedWeek != summary.weekStart) return 0
        return LiftrPreferences.weeklyTasksLastSeenCompletedCount(context)
    }

    fun markSeen(context: Context, summary: WeeklyTasksHomeSummary) {
        LiftrPreferences.setWeeklyTasksLastSeenCompletedCount(context, summary.completedCountThisWeek)
        LiftrPreferences.setWeeklyTasksLastSeenWeekStart(context, summary.weekStart)
    }

    fun accentDotColor(context: Context, summary: WeeklyTasksHomeSummary): Color? {
        if (hasUnseenCompletion(context, summary)) {
            return Color(0xFF34C759)
        }
        if (summary.hasPickableTasks) {
            return Color(0xFFFF9500)
        }
        return null
    }

    fun pillShowsPickSubtitle(summary: WeeklyTasksHomeSummary): Boolean = summary.hasPickableTasks
}

suspend fun fetchWeeklyTasksHomeSummary(supabase: io.github.jan.supabase.SupabaseClient): WeeklyTasksHomeSummary? =
    runCatching {
        val res = supabase.postgrest.rpc(BackendContracts.Rpc.GET_MY_WEEKLY_TASKS_HOME_SUMMARY_V1) { }
        val o = JSONObject(res.data)
        WeeklyTasksHomeSummary(
            weekStart = o.optString("week_start"),
            acceptedCount = o.optInt("accepted_count", 0),
            acceptSlotsRemaining = o.optInt("accept_slots_remaining", 0),
            availableCount = o.optInt("available_count", 0),
            completedCountThisWeek = o.optInt("completed_count_this_week", 0),
        )
    }.getOrNull()

suspend fun markWeeklyTasksHomeSeen(supabase: io.github.jan.supabase.SupabaseClient, context: Context) {
    val summary = fetchWeeklyTasksHomeSummary(supabase) ?: return
    WeeklyTasksHomeBadgeState.markSeen(context, summary)
}

File: Liftr/WeeklyTasksHomeBadgeState.swift
import Foundation
import SwiftUI

struct WeeklyTasksHomeSummary: Decodable, Equatable {
    let week_start: Date
    let accepted_count: Int
    let accept_slots_remaining: Int
    let available_count: Int
    let completed_count_this_week: Int

    var hasPickableTasks: Bool {
        accept_slots_remaining > 0 && available_count > 0
    }
}

enum WeeklyTasksHomeBadgeState {
    private static let lastSeenCountKey = "weeklyTasksLastSeenCompletedCount"
    private static let lastSeenWeekStartKey = "weeklyTasksLastSeenWeekStart"

    private static var defaults: UserDefaults { .standard }

    private static func weekStartKey(_ weekStart: Date) -> String {
        ISO8601DateFormatter().string(from: weekStart)
    }

    private static func effectiveLastSeenCount(for summary: WeeklyTasksHomeSummary) -> Int {
        let storedWeek = defaults.string(forKey: lastSeenWeekStartKey)
        let currentWeek = weekStartKey(summary.week_start)
        if storedWeek != currentWeek {
            return 0
        }
        return defaults.integer(forKey: lastSeenCountKey)
    }

    static func hasUnseenCompletion(_ summary: WeeklyTasksHomeSummary) -> Bool {
        summary.completed_count_this_week > effectiveLastSeenCount(for: summary)
    }

    static func markSeen(_ summary: WeeklyTasksHomeSummary) {
        defaults.set(summary.completed_count_this_week, forKey: lastSeenCountKey)
        defaults.set(weekStartKey(summary.week_start), forKey: lastSeenWeekStartKey)
    }

    static func accentDotColor(for summary: WeeklyTasksHomeSummary) -> Color? {
        if hasUnseenCompletion(summary) {
            return Color(red: 0.204, green: 0.780, blue: 0.349)
        }
        if summary.hasPickableTasks {
            return Color(red: 1.0, green: 0.584, blue: 0.0)
        }
        return nil
    }

    static func pillShowsPickSubtitle(for summary: WeeklyTasksHomeSummary) -> Bool {
        summary.hasPickableTasks
    }
}

enum WeeklyTasksHomeSummaryLoader {
    static func fetch() async -> WeeklyTasksHomeSummary? {
        do {
            let res = try await SupabaseManager.shared.client
                .rpc("get_my_weekly_tasks_home_summary_v1")
                .execute()
            return try JSONDecoder.supabase().decode(WeeklyTasksHomeSummary.self, from: res.data)
        } catch {
            return nil
        }
    }

    static func markSeenIfPossible() async {
        guard let summary = await fetch() else { return }
        WeeklyTasksHomeBadgeState.markSeen(summary)
    }
}
